fulltext/__init__.py

```diff
@@ -1,339 +1,166 @@
 
+        with tempfile.NamedTemporaryFile(dir=FULLTEXT_TEMP, suffix=ext) as t:
+            shutil.copyfileobj(f, t)
+            t.flush()
+            return backend._get_path(t.name, **kwargs)
 
-# A placeholder for a kwarg default value.
-class NoDefault(object):
-    pass
+    else:
+        BackendError('Backend must provide `_get_path()` or `_get_file()`')
 
 
-def get(f, default=NoDefault, filename=None, type=None):
+def get(path_or_file, default=SENTINAL, mime=None, name=None, **kwargs):
     """
-    Gets text from a given file. The first parameter can be a path or a file-like object that
-    has a read method. Default is a way to supress errors and just return the default text.
-    Filename can help figure out the type of file being used if you passed in a file-like
-    object. Type can be used to override the type guessing. Type should be a tuple: (mimetype, encoding)
-    as returned by the mimetypes library.
+    Get document full text.
 
-    For any recognizable type, a command will be execute to extract the text. Then that text
-    will be post processed to remove redundant whitespace.
+    Accepts a path or file-like object. If given, `default` is returned
+    instead of an error. `mime` and `name` should be passed if the information
+    is available to caller, otherwise a best guess is made.
+    """
+    if not name:
+        name = getattr(path_or_file, 'name', None)
 
-    For zip files, the first member is extracted and post processed.
+    if not name and isinstance(path_or_file, string_types):
+        name = basename(path_or_file)
 
-    Any file whose type cannot be determined will simply be read then post processed.
-    """
-    if not isinstance(f, basestring) and filename is None:
-        # Try to help figure out the file type.
-        filename = getattr(f, 'name', '')
+    try:
+        backend_name = kwargs.pop('backend')
+    except KeyError:
+        if name:
+            ext = splitext(name)[1].lstrip('.')
+        elif mime:
+            ext = mime.partition('/')[2]
+        else:
+            ext = ''
+        backend_name = ext.replace('-', '_').lower()
+
+    if backend_name not in BACKENDS:
+        LOGGER.warning('Falling back to text backend')
+        backend = BACKENDS['text']
     else:
-        # If user provided filename, don't override it.
-        if filename is None:
-            filename = f
-        if not os.path.exists(filename):
-            if default is not NoDefault:
-                return default
-            raise FullTextException('File not found')
-    if not filename:
-        try:
-            filename = os.path.basename(f.url)
-        except AttributeError:
-            pass
-    if type is None:
-        type = get_type(filename)
-    handler = FUNC_MAP.get(type, read_content)
+        backend = BACKENDS[backend_name]
+
     try:
-        text = handler(f, type)
-    except:
-        if default is not NoDefault:
+        if isinstance(path_or_file, string_types):
```

Comment:
Wasn't this already done on line 121 in this method?

Reply:
Yes.